.changeset/curvy-moles-applaud.md

```diff
@@ -0,0 +1,53 @@
+---
+'@xstate/react': minor
+---
+
+You can conveniently create a subscription to the interpreter inside `useInterpret` and `useMachine` hooks.
```

Comment:
I have some concerns related to this. A convenient solution might make it too easy for people to subscribe to the state like this - but that's usually something that should be avoided.

Could you mention what kind of real-life use cases motivate this proposal?

Reply:
TLDR;
It allows for an easier way to avoid having `useInterpret|useMachine` + `useEffect` combo for cases such as persisting the state to local storage on every state change, for example.

Reply:
I'm a little bit more pro-composition and in this case, creating a service and subscribing to it are two separate concerns (IMHO). This is especially annoying in React where we can't even use hooks conditionally - so for the majority of `subscribe`-less cases, we would still clobber the internal hooks slot which would in turn clobber the React devtools UI with a redundant item (not that this list isn't clobbered today though 🤣 ).

Like, even `react-redux` doesn't come with `useSelector` that would use `shallowEqual` as the comparator (even though that was most likely this was an expectation of most of their users at first). Their official response is a copy-pastable recipe:
https://react-redux.js.org/api/hooks#recipe-useshallowequalselector

When it comes to `useMachine` this feels like a super straightforward (minute the types, types are complicated) recipe:
```ts
function useSubscribableMachine(getMachine, { subscribe, ...options } = {}) {
    const result = useMachine(getMachine, options);
    useIsomorphicLayoutEffect(() => {
        if (subscribe) {
            return service.subscribe(subscribe).unsubscribe;
        }
    }, [subscribe]);
    return result;
}
```

And when it comes to `useInterpret`... this is already supported 😅 :
```ts
useInterpret(machine, {}, state => { /* do what u want */ })
```

I don't strongly mind this extension but I think that the "slim core" has a certain beauty to it. Extending low-level bits is what writing software is - if we need to add something on top of an existing block we can just extend it and use it throughout the app.

Reply:
I had no idea there's a 3rd argument and I believe most users don't know that either :D